packages/pyright-internal/src/common/uri/baseUri.ts

```diff
@@ -0,0 +1,288 @@
+
+    abstract matchesRegex(regex: RegExp): boolean;
+
+    replaceExtension(ext: string): Uri {
+        const dir = this.getDirectory();
+        const base = this.filename;
+        const newBase = base.slice(0, base.length - this.extname.length) + ext;
+        return dir.combinePaths(newBase);
+    }
+
+    addExtension(ext: string): Uri {
+        return this.addPath(ext);
+    }
+
+    abstract addPath(extra: string): Uri;
+
+    // Returns a URI where the path is the directory name of the original URI, similar to the UNIX dirname command.
+    getDirectory(): Uri {
+        return this.getDirectoryImpl();
+    }
+
+    getRootPathLength(): number {
+        return this.getRootPath().length;
+    }
+
+    // Determines whether a path consists only of a path root.
+    abstract isRoot(): boolean;
+
+    // Determines whether a Uri is a child of some parent Uri.
+    abstract isChild(parent: Uri, ignoreCase?: boolean): boolean;
+
+    abstract isLocal(): boolean;
+
+    isUntitled(): boolean {
+        return this.scheme === 'untitled';
+    }
+
+    equals(other: Uri | undefined, ignoreCase?: boolean): boolean {
+        if (ignoreCase) {
+            return this.key.toLowerCase() === other?.key.toLowerCase();
+        }
+        return this.key === other?.key;
+    }
+
+    abstract startsWith(other: Uri | undefined, ignoreCase?: boolean): boolean;
+
+    pathStartsWith(name: string): boolean {
+        // ignore path separators.
+        name = normalizeSlashes(name);
+        return this.getComparablePath().startsWith(name);
+    }
+
+    pathEndsWith(name: string): boolean {
+        // ignore path separators.
+        name = normalizeSlashes(name);
+        return this.getComparablePath().endsWith(name);
+    }
+
+    pathIncludes(include: string): boolean {
+        // ignore path separators.
+        include = normalizeSlashes(include);
+        return this.getComparablePath().includes(include);
+    }
+
+    // How long the path for this Uri is.
+    abstract getPathLength(): number;
+
+    // Combines paths to create a new Uri. Any '..' or '.' path components will be normalized.
+    combinePaths(...paths: string[]): Uri {
+        return this.combinePathsImpl(...paths);
+    }
+
+    getRelativePath(child: Uri): string | undefined {
+        if (this.scheme !== child.scheme) {
+            return undefined;
+        }
+
+        // Unlike getRelativePathComponents, this function should not return relative path
+        // markers for non children.
+        if (child.isChild(this, false)) {
+            const relativeToComponents = this.getRelativePathComponents(child);
+            if (relativeToComponents.length > 0) {
+                return ['.', ...relativeToComponents].join('/');
+            }
+        }
+        return undefined;
+    }
+
+    getPathComponents(): readonly string[] {
+        // Make sure to freeze the result so that it can't be modified.
+        return Object.freeze(this.getPathComponentsImpl());
```

Comment:
so, this one having `xxxImpl` make sense since there is common code applied to `xxxImpl`